### DISCUSSION TITLE
How to map EDC to IDS within an extension?

### DISCUSSION THREAD

**Discussion by carlos-schmidt** [eclipse-edc/Connector#685] [Q&A]
Hello,
from an extension point of view: 

- How do I map/transform EDC assets and contracts onto the IDS Infomodel (Resource, Artifact, Contract, ...)?

- In the folder /data-protocols/ids/ids-transform-v1 there are various of these transformers. How can/should they be used inside of an extension?

Thanks in advance

**Comment by juliapampus** [comment]
Hi @carlos-schmidt!

Please take a look at e.g. `org.eclipse.dataspaceconnector.ids.api.multipart.dispatcher.sender.MultipartContractAgreementSender` lines 99-103. You can use the transformer by handing over the object that should be transformed, and the class of the resulting object.

```
var transformationResult = getTransformerRegistry().transform(contractAgreement, ContractAgreement.class);
if (transformationResult.failed()) {
    throw new EdcException("Failed to create IDS contract agreement");
}
```

The right transformer will be automatically detected by input and output type. You can add own transformers by implementing `IdsTypeTransformer` and adding the new class to the list in `IdsTransformServiceExtension`. We currently provide some tranformers for converting IDS objects to EDC objects and vice versa. There, you can also find e.g. asset to IDS resource/representation/artifact and IDS resource/representation/artifact to asset.

**Reply by juliapampus** [reply]
In the `initialize` method of your extensions class, you need to create the `TransformerRegistry` and add in to the context. See an example in e.g. `org.eclipse.dataspaceconnector.ids.core.IdsCoreServiceExtension`.

**Reply by carlos-schmidt** [reply]
Hello,

thanks for the answer! 
In `org.eclipse.dataspaceconnector.ids.api.multipart.handler.description.ArtifactDescriptionRequestHandler`, an `Artifact`  is being fetched from the `AssetIndex`, and later transformed to an `artifact`.
This means, the requestor must know which (asset) ID belongs to what kind of IDS object type, right?

**Reply by DominikPinsel** [reply]
To make it easier to see whether an object belongs to the `edc` or `ids` domain I will add a prefix in this comment. There is no meaning beyond this.

> This means, the requestor must know which (asset) ID belongs to what kind of IDS object type, right?

In case of the IDS transformation the `edc:Asset` is always transformed into a triplet of 1x `ids:Resource`, 1x `ids:Representation` and 1x `ids:Artifact`. Therefore, it is not necessary to know the asset ID to transform it into the correct ids object

> How do I map/transform EDC assets and contracts onto the IDS Infomodel (Resource, Artifact, Contract, ...)?

The edc / ids mapping of contracts is probably a little bit more complicated. I will try to showcase it here using the contract offer generation. *(Maybe I don't get the relationship between all the classes 100% right.)*

Contract Offer Generation:

The `edc:ContractOffer` are generated on the fly and not persisted anywhere. When the connector is asked about its `ids:ContractOffers` it checks the `edc:ContractDefinitions`,  that contain the contract details and a `AssetSelectorExpression`. The `AssetSelectorExpression` can be used to query the `AssetIndex` for all the `edc:Assets`, the `edc:ContractOffer` should be generated for. For each `edc:Asset` a new `edc:ContractOffer` is created. The creation of `edc:ContractOffers` is be done by the `ContractOfferService`. 
The `edc:ContractOffer` are then converted to `ids:ContractOffer` using the `TransformerRegistry`.
